recirq/otoc/otoc_circuits.py

```diff
@@ -0,0 +1,499 @@
+            z_only=z_only, ancilla=ancilla, cycles_per_echo=cycles_per_echo,
+            reverse=True, light_cones=[light_cone_u_dagger, light_cone_meas],
+            echo_indices=echo_indices_back)
+
+    butterfly_ops = [None, cirq.X, cirq.Y, cirq.Z]
+    prep_phases = [0.0, 1.0]
+    meas_phases = [0.0, 1.0]
+    otoc_circuits = []
+
+    for b_ops in butterfly_ops:
+        for p_prep in prep_phases:
+            for p_meas in meas_phases:
+                init_ops = [cirq.PhasedXPowGate(
+                    phase_exponent=p_prep, exponent=0.5)(ancilla),
+                            cirq.Y(qubits[0]) ** 0.5]
+                circuit_full = cirq.Circuit(init_ops)
+
+                if use_physical_cz:
+                    circuit_full.append(cirq.CZ(ancilla, qubits[0]))
+                    additional_layer = []
+                else:
+                    cz_seq = cz_to_sqrt_iswap(ancilla, qubits[0],
+                                              corrections=cz_correction)
+                    circuit_full.append(cz_seq[:-1])
+                    additional_layer = cz_seq[-1:]
+
+                if z_only:
+                    for q in qubits[1:]:
+                        if (q.row + q.col) % 2 == 0:
+                            additional_layer.append(cirq.Y(q))
+                else:
+                    additional_layer.append(
+                        [cirq.Y(q) ** 0.5 for q in qubits[1:]])
+
+                circuit_full.append(additional_layer)
+                circuit_full.append(circuits_forward[0])
+                if padding is not None:
+                    moment_pad = cirq.Moment(
+                        [cirq.WaitGate(duration=cirq.Duration(
+                            nanos=padding))(q) for q in qubits])
+                    circuit_full.append(moment_pad)
+
+                if b_ops is not None:
+                    if multi_b is True:
+                        b_mom = cirq.Moment(
+                            [b_ops(q_b) for q_b in butterfly_qubits])
+                        circuit_full.append(b_mom)
+                    else:
+                        circuit_full.append(
+                            b_ops(butterfly_qubits),
+                            strategy=cirq.InsertStrategy.NEW)
+
+                circuit_full.append(circuits_backward[0])
+
+                if use_physical_cz:
+                    circuit_full.append(cirq.CZ(ancilla, qubits[0]))
+                else:
+                    circuit_full.append(cz_to_sqrt_iswap(
+                        ancilla, qubits[0], corrections=cz_correction))
+
+                circuit_full.append(cirq.PhasedXPowGate(
+                    phase_exponent=p_meas, exponent=0.5)(ancilla))
+                circuit_full.append(cirq.measure(ancilla, key='z'),
+                                    strategy=cirq.InsertStrategy.NEW)
+                otoc_circuits.append(circuit_full)
+    return otoc_circuits
+
+
+def replace_gates(
```

Comment:
This is a function that replaces a certain number of gates in a Clifford circuit with non-Clifford gates. It's not quite used here since iSWAP is not supported (sqrt(iSWAP) is not a Clifford so this function doesn't serve any purpose for circuits constructed with currently available gates. I will modify the doc-string to mention this fact.